compiler/p/codegen/OMRMemoryReference.cpp

```diff
@@ -1617,37 +1617,79 @@ void OMR::Power::MemoryReference::accessStaticItem(TR::Node *node, TR::SymbolRef
          symbol->getStaticSymbol()->setTOCIndex(tocIndex);
          }
 
-      // We need a snippet for unresolved or AOT if:
-      //  1. the load hasn't been resolved yet
-      //  (otherwise optimizer will remove the ResolveCHK and we can be sure pTOC will contain the resolved address),
-      //  2. we don't have a PTOC slot, we must always take the slow path
-
-      if ((ref->isUnresolved() || useUnresSnippetToAvoidRelo) &&
-          (topNode->getOpCodeValue() == TR::ResolveCHK || tocIndex == PTOC_FULL_INDEX))
+      if (cg->comp()->target().cpu.isAtLeast(OMR_PROCESSOR_PPC_P10))
          {
-         snippet = new (cg->trHeapMemory()) TR::UnresolvedDataSnippet(cg, node, ref, isStore, false);
-         cg->addSnippet(snippet);
-         }
+         TR::Register *addrReg = cg->allocateRegister();
+         _baseRegister = addrReg;
+         self()->setBaseModifiable();
 
-      // TODO: Improve the code sequence for cases when we know pTOC is full.
-      TR::MemoryReference *tocRef = new (cg->trHeapMemory()) TR::MemoryReference(cg->getTOCBaseRegister(), 0, sizeof(uintptr_t), cg);
-      tocRef->setSymbol(symbol, cg);
-      tocRef->getSymbolReference()->copyFlags(ref);
-      tocRef->setUsingStaticTOC();
-      if (snippet != NULL)
-         {
-         tocRef->setUnresolvedSnippet(snippet);
-         tocRef->adjustForResolution(cg);
+         // For now, P10 PC-relative loads and stores are not supported for UnresolvedDataSnippet
+         // and for anything requiring a TR_ClassAddress relocation. To make things work, we must
+         // emit the 5 instruction load address by faking that the pTOC was full.
+         if (ref->isUnresolved() || useUnresSnippetToAvoidRelo || (cg->comp()->compileRelocatableCode() && symbol->isStatic() && symbol->isClassObject()))
```

Comment:
are you worried about PicSite handling such that you special-case for isClassObject()? we can certainly make it work.

Reply:
As the comment above this code mentions, this is for loads which need to generate `TR_ClassAddress` relocations when performing non-SVM AOT compiles and for loads which will be patched by resolution code in `PicBuilder`. While it is possible to get these cases working in theory, getting `ConstantDataSnippet` slots needing relocations to work correctly under AOT or that will be patched by `PicBuilder` is outside the scope of this PR, so for now this code is needed to ensure that the old behaviour is still used.